### PR TITLE
contest-questions APIs

### PR DESCRIPTION
**Pull request by ridhishjain**
#### APIs:
* *`addQuestion`*
**`POST /contests/:contest_id/:questions` Moderator only**: Add question

* *`removeQuestion`*
**`DELETE /contests/:contest_id/questions` Moderator only**: remove question

* *`getAllQuestions`*
**`GET /contests/:contest_id/questions`**: accessible to all, if public contest otherwise to group members

* *`getQuestion`*
**`GET /contests/:contest_id/questions/:question_id`**: accessible to all, if public contest otherwise to group members

### CODE REVIEW

File: server/models/contests/getQuestion.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+function getQuestion({ username, params }) {
+  return new Promise((resolve, reject) => {
+    const { contest_id: contestId, question_id: questionId } = params
+    pool.query(
+      `SELECT question_id, max_score FROM contests_questions cq
+      INNER JOIN contests c ON c.id = cq.contest_id WHERE (cq.question_id = ?) 
+      AND ((c.public = 1) OR (EXISTS(SELECT 1 FROM contests_groups cg INNER JOIN user_groups ug ON cg.group_id = ug.group_id 
+        WHERE ug.username = ? AND cg.contest_id = ?)))`,
```
Comment (cjchirag7) [inline server/models/contests/getQuestion.js:17]:
```suggestion
      `SELECT q.name, q.difficulty, q.type, max_score, q.creator, q.problem_statement, q.input_format, q.output_format, q.constraints, q.options FROM contests_questions cq
      INNER JOIN contests c ON c.id = cq.contest_id
      INNER JOIN questions q ON cq.question_id=q.id
      WHERE (c.id=? AND q.id=?) AND (EXISTS(SELECT 1 from contests_moderators WHERE contest_id=? AND moderator=?) OR (NOW()>c.end_time AND c.confidential_questions=0 AND (c.public = 1
      OR EXISTS(SELECT 1 FROM contests_groups cg INNER JOIN user_groups ug ON cg.group_id = ug.group_id 
        WHERE ug.username = ? AND cg.contest_id = ?))) OR (NOW()>=c.start_time AND NOW()<=c.end_time AND EXISTS(SELECT 1 FROM contests_participants WHERE contest_id=? AND participant=?)))`,
```

File: server/models/contests/getQuestion.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      [questionId, username, contestId],
```
Comment (cjchirag7) [inline server/models/contests/getQuestion.js:18]:
```suggestion
      [contestId, questionId, contestId, username, username, contestId, contestId, username],
```

File: server/models/contests/getQuestion.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      (error, results) => {
+        if (error || results === undefined) {
+          return reject(error)
+        }
+        if (!results || !results.length) {
+          return reject(
+            'The contest is private and user do not belongs to any eligible groups'
```
Comment (cjchirag7) [inline server/models/contests/getQuestion.js:25]:
```suggestion
            'You do not have moderator access of the contest. If the contest is active, make sure that you are registered as a participant.'
```

File: server/models/contests/removeQuestion.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      `DELETE FROM contests_questions WHERE question_id = ? AND contest_id = (SELECT contest_id FROM contests_moderators WHERE contest_id=? AND moderator=?)`,
+      [questionId, contestId, username],
+      (error, res) => {
+        if (error || res === undefined) {
+          return reject(error)
+        } else {
+          const { affectedRows } = res
+          if (!affectedRows) {
+            return reject(
+              'Invalid question Id or the user do not have required permissions'
```
Comment (cjchirag7) [inline server/models/contests/removeQuestion.js:25]:
```suggestion
              'Invalid question Id or you do not have moderator access to the contest'
```

**Review by cjchirag7 — CHANGES_REQUESTED** [review]
While returning the questions of a contest, make sure that a normal user is not able to see the questions before the contest.

File: server/models/contests/getAllQuestions.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+function getAllQuestions({ username, params }) {
+  return new Promise((resolve, reject) => {
+    const { contest_id: contestId } = params
+    pool.query(
+      `SELECT question_id, max_score FROM contests_questions cq
+      INNER JOIN contests c ON c.id = cq.contest_id WHERE (c.public = 1)
+      OR (EXISTS(SELECT 1 FROM contests_groups cg INNER JOIN user_groups ug ON cg.group_id = ug.group_id 
+        WHERE ug.username = ? AND cg.contest_id = ?))`,
```
[the remark this answers is not in the document]
Comment (ridhishjain) [reply server/models/contests/getAllQuestions.js:17]: got it

**Review by cjchirag7 — APPROVED** [review]
LGTM!

**Comment by ridhishjain** [comment]
dont merge I'm still checking the queries

File: server/models/contests/removeQuestion.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ * @param {String} param0.username
+ * @return {Promise}
+ */
+
+function removeQuestion({ params, body, username }) {
+  return new Promise((resolve, reject) => {
+    const { contest_id: contestId } = params
+    const { question_id: questionId } = body
+    pool.query(
+      `DELETE FROM contests_questions WHERE question_id = ? AND contest_id = (SELECT contest_id FROM contests_moderators WHERE contest_id=? AND moderator=?)`,
```
Comment (NBNARADHYA) [inline server/models/contests/removeQuestion.js:16]: I think it is better to use either `EXISTS` or `JOIN` which execute much faster
```suggestion
      `DELETE FROM contests_questions WHERE (SELECT 1 FROM contests_moderators WHERE contest_id=? AND moderator=?) AND question_id = ? AND contest_id = ?`,
```

Comment (NBNARADHYA) [reply server/models/contests/removeQuestion.js:16]: Or use `JOIN` like this:
```suggestion
      `DELETE cq FROM contests_questions cq INNER JOIN contests_moderators cm ON cm.contest_id = cq.contest_id  WHERE cm.moderator = ? AND cq.question_id = ? AND contest_id = ?`,
```

File: server/models/contests/removeQuestion.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      [questionId, contestId, username],
```
Comment (NBNARADHYA) [inline server/models/contests/removeQuestion.js:17]: For `EXISTS`
```suggestion
      [contestId, username, questionId, contestId],
```

Comment (NBNARADHYA) [reply server/models/contests/removeQuestion.js:17]: For `JOIN`:
```suggestion
      [username, questionId, contestId],
```

**Review by NBNARADHYA — CHANGES_REQUESTED** [review]
Just some changes, everything else LGTM!

File: server/models/contests/removeQuestion.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ * @param {String} param0.username
+ * @return {Promise}
+ */
+
+function removeQuestion({ params, body, username }) {
+  return new Promise((resolve, reject) => {
+    const { contest_id: contestId } = params
+    const { question_id: questionId } = body
+    pool.query(
+      `DELETE FROM contests_questions WHERE (SELECT 1 FROM contests_moderators WHERE contest_id=? AND moderator=?) AND question_id = ? AND contest_id = ?`,
```
Comment (NBNARADHYA) [inline server/models/contests/removeQuestion.js:16]: I actually meant to say this: I forgot `EXISTS`, i have now updated
```suggestion
      `DELETE FROM contests_questions WHERE EXISTS(SELECT 1 FROM contests_moderators WHERE contest_id=? AND moderator=?) AND question_id = ? AND contest_id = ?`,
```

**Review by NBNARADHYA — CHANGES_REQUESTED** [review]

**Review by NBNARADHYA — APPROVED** [review]
LGTM!

**Pull request merged**